core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala

```diff
@@ -138,13 +154,22 @@ private[spark] class IndexShuffleBlockResolver(
       mapId: Int,
       lengths: Array[Long],
       dataTmp: File): Unit = {
+    val mapLocks = shuffleIdToLocks.get(shuffleId)
+    require(mapLocks != null, "Shuffle should be registered to IndexShuffleBlockResolver first")
+    val lock = mapLocks.synchronized {
```

Comment:
in most cases, you'll still have contention on this lock, right, its just a much smaller critical section?  As the usual case is for multiple tasks in the same executor to be for the same stage.

Reply:
The theory is many fewer threads would contend here because it's per-shuffleID.

If it's an issue, then your idea of a second-level ConcurrentHashMap might help. It's more complex than a usual Map but can allow for safe concurrent access by a limited number of threads.

Otherwise it might be overkill as the second-level Map.

Reply:
in the usual case, multiple threads are still sharing the same shuffleID, they're just writing to different map tasks.  (they are in the simple example job @ConeyLiu  gave, of `spark.range(0, 10000000, 1, 100).repartition(200).count()`)

Reply:
Oh I see, then your proposal could actually resolve the contention. With ConcurrentHashMap this synchronized block can be an updateIfAbsent call that might not even contend.